File: backend/src/AWSomeShop.API/Services/AuditLogService.cs
using AWSomeShop.API.Models.Entities;
using AWSomeShop.API.Repositories.Interfaces;
using AWSomeShop.API.Services.Interfaces;
using System.Text.Json;

namespace AWSomeShop.API.Services;

public class AuditLogService : IAuditLogService
{
    private readonly IAuditLogRepository _auditLogRepository;
    private readonly ILogger<AuditLogService> _logger;

    public AuditLogService(
        IAuditLogRepository auditLogRepository,
        ILogger<AuditLogService> logger)
    {
        _auditLogRepository = auditLogRepository;
        _logger = logger;
    }

    public async Task LogAsync(
        string adminId,
        string adminName,
        string action,
        string entityType,
        string? entityId = null,
        object? oldValue = null,
        object? newValue = null,
        string? ipAddress = null,
        string? userAgent = null)
    {
        var auditLog = new AuditLog
        {
            AdminId = adminId,
            AdminName = adminName,
            Action = action,
            EntityType = entityType,
            EntityId = entityId,
            OldValue = oldValue != null ? JsonSerializer.Serialize(oldValue) : null,
            NewValue = newValue != null ? JsonSerializer.Serialize(newValue) : null,
            IpAddress = ipAddress,
            UserAgent = userAgent,
            CreatedAt = DateTime.UtcNow
        };

        await _auditLogRepository.CreateAsync(auditLog);
        _logger.LogInformation("Audit log created: {Action} on {EntityType} by {AdminName}", action, entityType, adminName);
    }

    public async Task<(List<AuditLog> Logs, int TotalCount)> GetLogsAsync(
        int pageNumber = 1,
        int pageSize = 20,
        string? adminId = null,
        string? action = null,
        string? entityType = null,
        DateTime? startDate = null,
        DateTime? endDate = null)
    {
        return await _auditLogRepository.GetLogsAsync(
            pageNumber, pageSize, adminId, action, entityType, startDate, endDate);
    }

    public async Task<List<AuditLog>> GetLogsByEntityAsync(string entityType, string entityId)
    {
        return await _auditLogRepository.GetLogsByEntityAsync(entityType, entityId);
    }
}

File: frontend/src/services/auditLogService.ts
import { get } from '../utils/request';

export interface AuditLog {
  id: string;
  adminId: string;
  adminName: string;
  action: string;
  entityType: string;
  entityId?: string;
  oldValue?: string;
  newValue?: string;
  ipAddress?: string;
  userAgent?: string;
  createdAt: string;
}

export interface AuditLogListResponse {
  logs: AuditLog[];
  totalCount: number;
  page: number;
  pageSize: number;
  totalPages: number;
}

export interface AuditLogQueryParams {
  page?: number;
  pageSize?: number;
  adminId?: string;
  action?: string;
  entityType?: string;
  startDate?: string;
  endDate?: string;
}

/**
 * Admin: Get audit logs
 */
export const getAuditLogs = (params?: AuditLogQueryParams): Promise<AuditLogListResponse> => {
  return get<AuditLogListResponse>('/admin/audit-logs', params);
};

/**
 * Admin: Get audit logs for specific entity
 */
export const getAuditLogsByEntity = (entityType: string, entityId: string): Promise<{ logs: AuditLog[] }> => {
  return get<{ logs: AuditLog[] }>(`/admin/audit-logs/entity/${entityType}/${entityId}`);
};
